### PR TITLE
Support building for Magic Leap.

### PR DESCRIPTION
**Pull request by asajeffrey**
This PR gets Servo to build for Magic Leap, and provides a dummy Servo2D app.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because it's a port to a new architecture

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21985)

### CODE REVIEW

**Comment by asajeffrey** [comment]
It depends on https://github.com/servo/mozjs/pull/158 and https://github.com/servo/media/pull/158.

**Comment by asajeffrey** [comment]
Filed #21999 to get servo-media updated.

File: support/magicleap/Servo2D/code/src/Servo2D.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+// Destroy a Servo 2D instance
+Servo2D::~Servo2D() {
+  ML_LOG(Debug, "Servo2D Destructor.");
+  discard_servo(servo_);
+  servo_ = nullptr;
+}
+
+// The prism dimensions
+const glm::vec3 Servo2D::getInitialPrismExtents() const {
+  return glm::vec3(0.5f, 0.5f, 0.5f);
```
Comment (jdm) [inline support/magicleap/Servo2D/code/src/Servo2D.cpp]: Do these relate to the 500s in libmlservo, or is that just coincidence?

File: support/magicleap/Servo2D/code/src/Servo2D.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+int Servo2D::init() {
+
+  ML_LOG(Debug, "Servo2D Initializing.");
+
+  // Set up the prism
+  createInitialPrism();
+  lumin::ui::Cursor::SetScale(prism_, 0.03f);
+  instanceInitialScenes();
+
+  // Get the panar resource that holds the EGL context
```
Comment (jdm) [inline support/magicleap/Servo2D/code/src/Servo2D.cpp]: panar?

File: support/magicleap/Servo2D/code/src/Servo2D.cpp [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+  }
+
+  content_node->setRenderResource(plane_id);
+
+  // Get the EGL context, surface and display.
+  EGLContext ctx = plane_->getEGLContext();
+  EGLSurface surf = plane_->getEGLSurface();
+  EGLDisplay dpy = eglGetDisplay(EGL_DEFAULT_DISPLAY);
+  eglMakeCurrent(dpy, surf, surf, ctx);
+  glViewport(0, 0, 500, 500);
```
Comment (jdm) [inline support/magicleap/Servo2D/code/src/Servo2D.cpp]: Does this need to match the 500s in libmlservo? Can we pass this value in to Rust?

File: support/magicleap/Servo2D/code/src/main.cpp [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <Servo2D.h>
+#include <ml_logging.h>
+
+int main(int argc, char **argv)
+{
+  ML_LOG(Debug, "Servo2D Starting.");
+  Servo2D myApp;
```
Comment (jdm) [inline support/magicleap/Servo2D/code/src/main.cpp]: Does argv contain meaningful values for us? Can we accept a URL on the command line?

File: support/magicleap/Servo2D/code/src/Servo2D.cpp [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  }
+}
+
+bool Servo2D::updateLoop(float fDelta) {
+  // Get the EGL context, surface and display.
+  EGLContext ctx = plane_->getEGLContext();
+  EGLSurface surf = plane_->getEGLSurface();
+  EGLDisplay dpy = eglGetDisplay(EGL_DEFAULT_DISPLAY);
+  eglMakeCurrent(dpy, surf, surf, ctx);
+  glViewport(0, 0, 500, 500);
```
Comment (jdm) [inline support/magicleap/Servo2D/code/src/Servo2D.cpp]: Let's make this a named constant somewhere.

**Review by jdm — CHANGES_REQUESTED** [review]

**Comment by asajeffrey** [comment]
I did some fixes, but can't test the C++ code because I'm at home rather than in the office.

**Comment by asajeffrey** [comment]
Rebased.

**Comment by asajeffrey** [comment]
Updated `mozjs_sys` which was the last external blocker.

**Comment by paulrouget** [comment]
I see that you introduced a new port. Did you consider using the c api of libsimpleservo instead of libmlservo?

**Comment by asajeffrey** [comment]
@paulrouget yes, originally I was using libsimpleservo but ran into horrible linking problems which went away if I used libservo directly. My suspicion is that the ML is going to be a different enough target from Android that we're going to need a different port at some point anyway.

**Comment by paulrouget** [comment]
libsimpleservo can be compiled without any android stuff (it compiles on Desktop for example). It's supposed to be capable of being dynamically loaded like libservo. I'd be interested to know what kind of issues you ran into.

I don't want to block this, but maybe later we could look at how to re-use libsimpleservo for ML.

**Comment by asajeffrey** [comment]
It's been a while, but there was some issue about `libsimpleservo` being dynamically loaded causing symbols not to be found. I may have fixed that in the mean time, e.g. by statically loading openssl.

File: support/magicleap/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+Build the mlservo library:
+```
+MAGICLEAP_SDK=*directory*  ./mach build -d --magicleap
+```
+This builds a static library `target/aarch64-linux-android/debug/libmlservo.a`.
+
+## Building the Servo2D application
+
+From inside the `ports/magicleap/Servo2D` directory:
```
Comment (ferjm) [inline support/magicleap/README.md]: s/ports/support

Comment (ferjm) [inline support/magicleap/README.md]: s/ports/support

**Comment by jdm** [comment]
@bors-servo delegate+
I'm fine with keeping the question about the URL as a followup.

**Pull request merged**